### PR TITLE
C51-76: Integrate Recent communications and Tasks on case summary tab

### PR DESCRIPTION
**Pull request by swastikpareek**
## Overview 
 As a part of this PR Recent Communications and Tasks are integrated to the app.

## After
c51-76 - after 1
c51-76 - after 2
c51-76 - after 4
c51-76 - after loading

## Technical Details
* Fix JS issue when clicking on empty next Activity and Next Milestone card.
* Add click to open in modal functionality to long  activity cards.
* Check if Bulk actions are available are now added as a service. Test added as well.
* New class `civicase__summary-tab-list-view` added for specific styling of recent comms and tasks.
* Fix next activity on the case summary tab
  ### Problem.
  The next activity was being calculated from the list of all activities. Since `allActivities` show only 25 results at a time it was not possible to get the next activity which is incomplete from the whole long list of activities for a case w/o getting all the activities which can be done by setting `limit:0`. If this is applied to `api.activity.get.1` options params then it might break the current functionality. Also the results will be too huge and will hinder performance. 
  ### Solution
  So the good solution for this was to have another chained API call with setting `limit:0` and contextually loading activities which are not completed by passing `status_id: {'!=': 'Completed'},` as a parameter to the get request.

  Creating a new chain call fixed the issue and also saved much on performance.
* Difference in click behaviour of `activityCard--long` on activity feed and non activity feed fixed.
  ### Problem.
    We are using `activityCard--long` on activity feed and at other places like (Recent Comms and Tasks and files tab). On activity feed we need to have different function to be called when clicking on the activity tile (Open detail view) and different behaviour on non activity feed type pages (Open in popup). We have different functions for it but we need to make the directive flexible to adjust for both the behaviours
  ### Fix
   A new attr `feedViewFn` solved the issue. So on click, if `feedViewFn` is present then it will be called otherwise default behaviour (open in popup) function `viewInPopup` function will be called. 
  ```html
  <div
    class="civicase__activity-card civicase__activity-card--long"
    ng-click="feedViewFn ? feedViewFn(activity.id, $event) : viewInPopup($event, activity)"
    ng-class="{
      'civicase__activity-card--draft': activity.status==='Draft',
      'civicase__activity-card--ribbon': activity.category.indexOf('milestone') > -1,
      'civicase__activity-card--with-checkbox': activity.category.indexOf('task') > -1,
      'civicase__activity-card--alert': activity.category.indexOf('alert') > -1,
      'civicase__activity-card--file': activity.type === 'File Upload'
    }">
   ...
   ...
  </div>
  ```
* Adds functionality to `Add a new task`  and `Add a new communication` button using directive `civicase-add-activity-menu`.

### CODE REVIEW

**Comment by deb1990** [comment]
2018-10-04 at 11 09 am
+ icon is missing from both task and recent communication empty state.

File: ang/civicase/Utils.js [hunk 1]
```diff
@@ -293,40 +293,65 @@
...
+        'contacts', 'start_date', 'end_date', 'is_deleted', 'activity_summary',
+        'activity_count', 'category_count', 'tag_id.name', 'tag_id.color',
+        'tag_id.description', 'tag_id.parent_id', 'related_case_ids'
+      ];
+      var caseListReturnParams = ['case_type_id', 'start_date', 'end_date', 'status_id', 'contacts', 'subject'];
+      var customValuesReturnParams = [
+        'custom_group.id', 'custom_group.name', 'custom_group.title',
+        'custom_field.name', 'custom_field.label', 'custom_value.display'
+      ];
+      var relationhipReturnParams = ['id', 'relationship_type_id', 'contact_id_a', 'contact_id_b', 'description', 'start_date'];
```
Comment (deb1990) [inline ang/civicase/Utils.js]: Typo, should be `relationshipReturnParams `

File: ang/civicase/Utils.js [hunk 2]
```diff
@@ -345,19 +370,29 @@
...
           'status_id.filter': 0,
           options: {limit: panelLimit, sort: 'activity_date_time ASC'},
-          return: ['activity_type_id', 'subject', 'activity_date_time', 'status_id', 'target_contact_name', 'assignee_contact_name', 'is_overdue', 'is_star', 'file_id', 'tag_id.name', 'tag_id.description', 'tag_id.color', 'file_id']
+          return: activityReturnParams
+        },
+        // For the "Next Activity" panel
+        'api.Activity.get.4': {
+          case_id: '$value.id',
+          status_id: {'!=': 'Completed'},
+          options: {
```
Comment (deb1990) [inline ang/civicase/Utils.js]: Why dont you send a filter param so that it only returns non milestone values. that way you can just limit the value to `1`. Instead of getting all.
Some thing like `'activity_type_id.grouping': {NOT LIKE: '%milestone%'},`. That way we dont have to do anything on the frontend to filter.

File: ang/test/civicase/CaseDetails.spec.js [hunk 3]
```diff
@@ -64,13 +64,14 @@ describe('civicaseCaseDetails', function () {
       compileDirective();
       element.isolateScope().item = CasesData.values[0];
       element.isolateScope().item.allActivities = CasesData.values[0]['api.Activity.get.1'].values;
-
+      element.isolateScope().item.nextActivities = CasesData.values[0]['api.Activity.get.4'].values;
       element.isolateScope().pushCaseData(CasesData.values[0]);
     });
 
-    it('calculates the incomplete tasks and scheduled activities', function () {
+    it('calculates the incomplete tasks and scheduled activities and nextActivitiyWhichIsNotMilestone', function () {
```
Comment (deb1990) [inline ang/test/civicase/CaseDetails.spec.js]: Please have 3 separate `it` for each of the `expect` statements.

File: ang/test/civicase/CaseDetails.spec.js [hunk 4]
```diff
@@ -64,13 +64,14 @@ describe('civicaseCaseDetails', function () {
...
       expect(element.isolateScope().item.category_count.scheduled).toEqual(getScheduledActivitiesCount(element.isolateScope().item.allActivities));
       expect(element.isolateScope().item.category_count.incomplete.task).toBe(2);
+      expect(element.isolateScope().item.nextActivityNotMilestone.id).toBe('1774');
```
Comment (deb1990) [inline ang/test/civicase/CaseDetails.spec.js]: How we do know this is `1774` exactly? we need to find the correct value.

File: ang/civicase/ActivityCard--Long.html [hunk 5]
```diff
@@ -1,6 +1,6 @@
 <div
   class="civicase__activity-card civicase__activity-card--long"
-  ng-click="viewInPopup($event, activity)"
+  ng-click="feedViewFn ? feedViewFn(activity.id, $event) : viewInPopup($event, activity)"
```
Comment (deb1990) [inline ang/civicase/ActivityCard--Long.html]: Why is this necessary? The click event should be mentioned outside of the card. For example
```
  <div
              case-activity-card="activity"
              mode="long"
              refresh-on-checkbox-toggle="true"
              bulk-allowed="bulkAllowed"
              ng-click="viewActivity(activity.id, $event)"
              refresh-callback="refreshCase">
            </div>
```

**Comment by swastikpareek** [comment]
@deb1990 : As far as I remember. We don't have `+` icon in the designs
no icon

File: ang/civicase/ActivityCard--Long.html (hunk 5 above)
Comment (swastikpareek) [reply ang/civicase/ActivityCard--Long.html]: `viewActivity` is of the outside the activity scope so it can be called **on** the card. But `viewInPopup` is inside the activity card scope so can't be called inside but need to be called inside. 
Now if we call the `viewInPopup ` inside then it will break outer ng click call and vice versa. So had this implementation.

File: ang/test/civicase/CaseDetails.spec.js (hunk 3 above)
Comment (swastikpareek) [reply ang/test/civicase/CaseDetails.spec.js]: Since the next activity is coming from the API response we don't need to write tests for it.

File: scss/components/_panel-transparent-header.scss [hunk 6]
```diff
@@ -16,9 +16,9 @@
       margin: 0 5px;   
     }
 
-    .panel-title-inner {
-      margin: 0 10px 0 0;
-    }
+    .civicase__case-card__activity-count:nth-child(2) {
```
Comment (deb1990) [inline scss/components/_panel-transparent-header.scss]: This should not be a part of `_panel-transparent-header.scss`. It should be mentioned in the context of `Tasks Panel`

File: scss/components/_panel-transparent-header.scss [hunk 7]
```diff
@@ -16,9 +16,9 @@
...
+      margin-left: 10px;
+    }
```
Comment (deb1990) [inline scss/components/_panel-transparent-header.scss]: Remove  extra spaces added.

File: ang/civicase/ActivityCard--Long.html [hunk 8]
```diff
@@ -1,6 +1,6 @@
 <div
   class="civicase__activity-card civicase__activity-card--long"
-  ng-click="feedViewFn ? feedViewFn(activity.id, $event) : viewInPopup($event, activity)"
+  ng-click="viewInPopup($event, activity)"
```
Comment (deb1990) [inline ang/civicase/ActivityCard--Long.html]: For every Activity Card, we should write `viewActivityDetails($event)`.
Inside the Activity card.
```
$scope.viewActivityDetails = function () {
  $scope.$emit();

  if ($scope.viewActivityInPopup) {
     $scope.viewInPopup();
  }
}
```

File: ang/civicase/ActivityCard.js [hunk 9]
```diff
@@ -24,16 +24,17 @@
...
         editActivityUrl: '=?editActivityUrl',
         type: '=type',
-        feedViewFn: '=feedViewFn'
+        viewActivityInFeed: '=?'
       }
     };
   });
 
-  function caseActivityCardController ($scope, getActivityFeedUrl, dialogService, templateExists, crmApi, crmBlocker, crmStatus, DateHelper) {
+  function caseActivityCardController ($scope, $rootScope, getActivityFeedUrl, dialogService, templateExists, crmApi, crmBlocker, crmStatus, DateHelper) {
```
Comment (deb1990) [inline ang/civicase/ActivityCard.js]: No need to include `$rootScope`. you can use `$scope.emit`

File: ang/civicase/ActivityCard.js [hunk 10]
```diff
@@ -24,16 +24,17 @@
         bulkAllowed: '=',
         editActivityUrl: '=?editActivityUrl',
         type: '=type',
-        feedViewFn: '=feedViewFn'
+        viewActivityInFeed: '=?'
```
Comment (deb1990) [inline ang/civicase/ActivityCard.js]: As discussed, the popup should not be the default behaviour. only when `viewActivityInPopup` is sent, we should call that function.
We should always send the emit.
But incase the `viewActivityInPopup` is true. We call function to open the popup.
